### PR TITLE
Improve transaction analysis (RIPD-755)

### PR DESCRIPTION
**Pull request by nbougalis**
The code which analyzes differences between the ledger we built and the ledger the network built can now perform more intelligent analysis to understand the differences. Specifically it can categorize:

* Differences in the result of the operations;
* Differences in the transaction ordering; and
* Differences in the generated metadata

Reviews @HowardHinnant, @JoelKatz

### CODE REVIEW

**Comment by HowardHinnant** [comment]
:+1: I like it.  Nice job!

**Comment by nbougalis** [comment]
David has seen this code prior to the pull request but he doesn't count as a reviewer. I'm adding @scottschurr as a second reviewer.

File: src/ripple/app/ledger/LedgerHistory.cpp [hunk 1]
```diff
@@ -158,26 +159,114 @@ log_metadata_difference(Ledger::pointer builtLedger, Ledger::pointer validLedger
...
     if (validMetaData != nullptr && builtMetaData != nullptr)
     {
-        WriteLog (lsERROR, LedgerMaster) <<
-            "MISMATCH tx differ in metadata only " << tx <<
-            " : built metadata is " << builtMetaData->getJson(0) <<
-            " : valid metadata is " << validMetaData->getJson(0);
+        auto const& validNodes = validMetaData->getNodes ();
+        auto const& builtNodes = builtMetaData->getNodes ();
+
+        bool result_diff =
```
Comment (scottschurr) [inline src/ripple/app/ledger/LedgerHistory.cpp]: Could make `result_diff` `const` like `index_diff` and `nodes_diff`.  I'd like that.

File: src/ripple/app/ledger/LedgerHistory.cpp [hunk 2]
```diff
@@ -158,26 +159,114 @@ log_metadata_difference(Ledger::pointer builtLedger, Ledger::pointer validLedger
...
+                    " Index: " << validMetaData->getIndex ();
+            }
+            else if (result_diff)
+            {
+                WriteLog (lsERROR, LedgerMaster) << "MISMATCH on TX " << tx <<
+                    ": Different result!";
+                WriteLog (lsERROR, LedgerMaster) << " Built:" <<
+                    " Result: " << builtMetaData->getResult ();
+                WriteLog (lsERROR, LedgerMaster) << " Valid:" <<
+                    " Index: " << validMetaData->getResult ();
```
Comment (scottschurr) [inline src/ripple/app/ledger/LedgerHistory.cpp]: I'm guessing " Index: " wants to be " Result: " here.

File: src/ripple/app/ledger/LedgerHistory.cpp [hunk 3]
```diff
@@ -158,26 +159,114 @@ log_metadata_difference(Ledger::pointer builtLedger, Ledger::pointer validLedger
...
+            }
+        }
+        else
+        {
+            if (result_diff && index_diff)
+            {
+                WriteLog (lsERROR, LedgerMaster) << "MISMATCH on TX " << tx <<
+                    ": Different result, index and nodes!";
+                WriteLog (lsERROR, LedgerMaster) << " Built:\n" <<
+                    builtMetaData->getIndex ();
```
Comment (scottschurr) [inline src/ripple/app/ledger/LedgerHistory.cpp]: `WriteLog` formatting is a bit different for this case compared to the others.  For consistency consider adding `" Index: " << xxxxxMetaData...` to lines 222 and 224.

**Comment by scottschurr** [comment]
The logic all looks sound.  The only questions of (minor) substance involve the precise text going to the log.  Once Nik addresses or comments on those I'd say it'll be good to go.

File: src/ripple/app/ledger/LedgerHistory.cpp (hunk 3 above)
Comment (nbougalis) [inline src/ripple/app/ledger/LedgerHistory.cpp]: Great catch!

File: src/ripple/app/ledger/LedgerHistory.cpp (hunk 2 above)
Comment (nbougalis) [inline src/ripple/app/ledger/LedgerHistory.cpp]: It does.

File: src/ripple/app/ledger/LedgerHistory.cpp (hunk 1 above)
Comment (nbougalis) [inline src/ripple/app/ledger/LedgerHistory.cpp]: Wilco

**Comment by scottschurr** [comment]
:+1:

**Comment by nbougalis** [comment]
I see what you mean. In this case we are not logging the nodes but the entire object. I guess I could make it look the similar.

**Comment by nbougalis** [comment]
Merged as 64259c7bcb3049d26695032db5b0638fa8cf5538

**Pull request closed without merging**